### PR TITLE
[CALCITE-4421] Throw IllegalArgumentException rather than AssertionError if the size of two lists to be combined is not equal

### PR DESCRIPTION
**Pull request by chunweilei**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/util/Pair.java [hunk 1]
```diff
@@ -185,7 +185,8 @@ public Pair(T1 left, T2 right) {
     final int size;
     if (strict) {
       if (ks.size() != vs.size()) {
-        throw new AssertionError();
+        throw new IllegalArgumentException(
+            "The size of two lists to be combined is required to be equal.");
```
Comment (liyafan82) [inline core/src/main/java/org/apache/calcite/util/Pair.java:189]: size -> sizes
is -> are

File: core/src/test/java/org/apache/calcite/util/UtilTest.java [hunk 2]
```diff
@@ -967,6 +968,17 @@ private String toString(byte[] bytes) {
...
 
+  /**
+   * Unit test for {@link Pair#zip(java.util.List, java.util.List, boolean)}.
+   */
+  @Test void testPairZipWithStrict() {
+    List<String> strings = Arrays.asList("paul", "george", "john", "ringo");
+    List<Integer> integers = Arrays.asList(1942, 1943);
+    assertThrows(IllegalArgumentException.class, () -> {
+      List<Pair<String, Integer>> zip = Pair.zip(strings, integers, true);
+    });
```
Comment (liyafan82) [inline core/src/test/java/org/apache/calcite/util/UtilTest.java:979]: Maybe we should hold the return value of `assertThrows`, and asserts the error message is as expected?

File: core/src/main/java/org/apache/calcite/util/Pair.java (hunk 1 above)
Comment (vlsi) [reply core/src/main/java/org/apache/calcite/util/Pair.java:189]: Should the sizes be included in the message as well?

**Comment by julianhyde** [comment]
-1 on this change.

As I explained in the JIRA case, it is an internal method, it doesn't need detailed messages.

The real fix is to createCastRel.

**Comment by liyafan82** [comment]
AFAIK, Calcite utilities are widely used in client code, because they are implemented with high quality. 
So I think the error message would benefit Calcite users.

**Comment by chunweilei** [comment]
> -1 on this change.
> 
> As I explained in the JIRA case, it is an internal method, it doesn't need detailed messages.
> 
> The real fix is to createCastRel.

@julianhyde, could you please review [https://github.com/chunweilei/calcite/commit/a0ccf9d13da2f47cfacd7dda8a5b9e51f382f2b8] to see whether it is your preference?

**Comment by chunweilei** [comment]
Close the PR. You can find more detail in https://issues.apache.org/jira/browse/CALCITE-4421.

**Pull request closed without merging**